### PR TITLE
[Live bidding] Fix edge cases, add and improve tests

### PR DESCRIPTION
**Pull request by vguerci**

### CODE REVIEW

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m [hunk 1]
```diff
@@ -253,77 +266,388 @@ - (void)testGetBid_GivenLiveBiddingIsDisabled_ThenGetBidThenFetch {
...
+  self.cacheManager.bidCache[self.adUnit1] = self.bid1;
+  [self givenApiHandlerRespondValidBid];
+
+  [self fetchLiveBidForAdUnit:self.adUnit1
+             andExpectCdbCall:NO
+                    bidCached:nil
+                  bidConsumed:nil
+                 bidResponded:nil];
 
   // Silent bid from cdb call has been cached
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:577]: The comment does not represent the expectation.
Isn't the expectation: silenced is not consumed from cache ?

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m [hunk 2]
```diff
@@ -333,14 +657,7 @@ - (void)testLiveBid_GivenExpiredBid_ThenNoResponseGiven {
   CR_CdbBid *expiredBid = CR_CdbBidBuilder.new.adUnit(self.adUnit1).expired().build;
   [self givenApiHandlerRespondBid:expiredBid];
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:658]: Is it normal to consider expiration when bid is not put in the cache ?

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m [hunk 3]
```diff
@@ -350,65 +667,177 @@ - (void)testLiveBid_GivenNoBid_ThenNoResponseGiven {
...
+  [self givenApiHandlerRespondBid:noBid];
+  [self givenTimeBudgetExceeded];
+
+  [self fetchLiveBidAndExpectBidCached:noBid bidConsumedAndResponded:self.bid1];
 
   // No bid from cdb call has not been cached
   XCTAssertNotEqual(self.cacheManager.bidCache[self.adUnit1], noBid);
 }
 
+- (void)testLiveBid_GivenConcurrentCalls_ThenBidsFromResponsesGivenWithOrder {
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:687]: Nit: the order is not important. The important is that concurrency exists.
With cache bidding, an ad unit can only be fetched once at the same time. With live bidding, there is no limitation.

**Comment by TurpIF** [comment]
LGTM, I just want to get your confirmation/infirmation on few points.

File: CriteoPublisherSdk/Sources/CR_BidManager.m [hunk 4]
```diff
@@ -182,35 +199,36 @@ - (void)fetchLiveBidForAdUnit:(CR_CacheAdUnit *)adUnit
                        CR_CdbBid *bid = cdbResponse.cdbBids[0];
                        if (bid.isInSilenceMode && !bid.isExpired) {
```
[the remark this answers is not in the document]
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:200]: Indeed, that is taken straight from bid caching and does not makes sense there, will fix, nice catch

File: CriteoPublisherSdk/Sources/CR_BidManager.m [hunk 5]
```diff
@@ -316,4 +334,25 @@ - (CR_Config *)config {
...
 }
 
+- (BOOL)isSlotSilent:(CR_CacheAdUnit *)adUnit {
+  BOOL silenced = NO;
+  CR_CdbBid *bid = [cacheManager getBidForAdUnit:adUnit];
+  if (bid.isInSilenceMode) {
+    if (!bid.isExpired) {
+      silenced = YES;
+    } else {
+      [self unsilenceSlotForAdUnit:adUnit];
```
[the remark this answers is not in the document]
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:344]: ¯\\\_(ツ)_/¯

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m [hunk 6]
```diff
@@ -253,77 +266,388 @@ - (void)testGetBid_GivenLiveBiddingIsDisabled_ThenGetBidThenFetch {
 }
 
 #pragma mark - Live Bidding
+#pragma mark Helpers
+
+- (void)expectBidConsumed:(CR_CdbBid *)bid {
+  if (bid) {
+    OCMExpect([self.feedbackDelegateMock onBidConsumed:bid]);
```
[the remark this answers is not in the document]
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:273]: Yup but that is not true with Reject I think? Which the else does, reason of this approach.

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 1 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:577]: This is a leftover of the test before, will edit the comment. Also this is a bit redundant with the bidConsumed expect above.

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 2 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:658]: Earlier test ported from bid caching that is related to that wrong expectation that a bid could be returned expired from CDB, will remove this test (or tests?) while fixing your other comment

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 3 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:687]: Ok, the order there is used to check that concurrency is properly handled and not FIFO, an interpretation of "at the same time" and to prevent queueing issues.

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (TurpIF) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:273]: You can use the `OCMVerify(never(), [mock doStuff]);`

File: CriteoPublisherSdk/Sources/CR_BidManager.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:200]: Done

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 1 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:577]: Done

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 2 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:658]: Done

**Review by TurpIF — APPROVED** [review]

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:273]: Still somewhat similar, expect before doing, but noted, thanks!

**Comment by vguerci** [comment]
Thanks!
